## Project Stage - III (Distributions and Hypothesis Testing)

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import matplotlib.pyplot as plt
import scipy.stats as stats
import statsmodels.formula.api as smf

## Task 2: Hypothesis Testing and Regression (50 pts)¶

### M2.1 Formulate Hypothesis for 5 identified variables in Stage 1 and test the hypothesis (25 pts)

##### The  variables I am considering for hypothesis are as follows: 
     *Drug overdose deaths raw value  
     *Injury deaths raw value
     *Adult smoking raw value
     *Firearm fatalities raw value
     *Insufficient sleep raw value
     


#### Formally state the Null and Alternative Hypothesis and Define the type of hypothesis and the thresholds

Hypothesis is supposition or proposed explanation made on the basis of limited evidence as a starting point for further investigation. Also, its necessary to test this to carry out the study of a sample to variables. Thus NUll hyporthesis come into picture. 

Null Hypothsis (H0) : It states that there is no significant difference between the specified population. 
Alternative Hypothesis (H1) : The alternative is opposite of null hypothesis stating that there is relation ship between the variables.

The types of hypothesis are - 
1. One tailed Test - performed to check if the sample mean is less than or greater than the population mean.
2. Two tailed test - performed to find the significant difference between the variables.

There are different threshold that will help us understand if a null hypothesis should be accepted or rejected like pvalue

if the pvalue is less than 0.05 then we reject the null hypothesis. 

There are other functions like chi-contigency to check relation between two variables that are independant. 
P-Paired Test to check the if the data of interest differs with two different categories like checking weight of a person before the disease attacked him and then checking the weight after recovering. The Null hypothesis here is the weight remains same before and after. 

With the example above we will try to check for other 5 variables. 

#### Conduct the test and discuss the results for all 5 variables

In [2]:
super_df = pd.read_csv("../../../../data/stage_3/super_df.csv")

In [3]:
hypo_df = super_df[['County' ,'Norm_deaths' ,'Drug overdose deaths raw value', 'Premature death raw value', 'Injury deaths raw value', 'Adult smoking raw value' , 'Firearm fatalities raw value' , 'Insufficient sleep raw value']]

In [4]:
hypo_df[["County", "State"]] = hypo_df["County"].str.split(",",expand = True)
hypo_df["State"] = hypo_df["State"].str.lstrip()

C:\Users\p_bud\AppData\Local\Temp\ipykernel_4268\3701686775.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hypo_df[["County", "State"]] = hypo_df["County"].str.split(",",expand = True)
C:\Users\p_bud\AppData\Local\Temp\ipykernel_4268\3701686775.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hypo_df[["County", "State"]] = hypo_df["County"].str.split(",",expand = True)
C:\Users\p_bud\AppData\Local\Temp\ipykernel_4268\3701686775.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of 

In [5]:
hypo_df.head(5)

County  Norm_deaths  Drug overdose deaths raw value  \
0  Abbeville County          7.0                       17.443577   
1     Acadia Parish         19.0                       13.843631   
2   Accomack County         11.0                       11.171482   
3        Ada County         12.0                       13.706962   
4      Adair County         11.0                             NaN   

   Premature death raw value  Injury deaths raw value  \
0                9517.898548                83.535478   
1               10194.322670                79.039494   
2               10459.767290                82.000073   
3                5027.391209                59.693597   
4                9140.796987                88.793248   

   Adult smoking raw value  Firearm fatalities raw value  \
0                 0.174334                          23.3   
1                 0.210474                          21.1   
2                 0.164887                          18.2   
3                 0.117813                          11.7   
4                 0.232604                          11.5   

   Insufficient sleep raw value State  
0                      0.360838    SC  
1                      0.323577    LA  
2                      0.367502    VA  
3                      0.262969    ID  
4                      0.364091    KY

#### I grouped the data by state and created a dataframe that gives the mean of each state for each variable. Using the mean we will compare with the deaths value to understand if the mean of the two samples are same. Here am performing the test of independance. 

In [6]:
df_mean_variables = hypo_df.groupby(['State'])[['Norm_deaths' ,'Drug overdose deaths raw value', 'Premature death raw value', 'Injury deaths raw value', 'Adult smoking raw value' , 'Firearm fatalities raw value' , 'Insufficient sleep raw value']].mean().sort_values(by  = "Norm_deaths", ascending = False).reset_index()

In [7]:
df_mean_variables

State  Norm_deaths  Drug overdose deaths raw value  \
0     WV    22.094340                       45.771399   
1     DC    22.000000                       34.385578   
2     NM    20.466667                       25.845132   
3     KY    19.043103                       32.039178   
4     DE    18.333333                       27.365376   
5     MD    17.666667                       31.606481   
6     TN    17.244681                       25.215544   
7     MA    17.000000                       29.212216   
8     RI    16.600000                       26.010388   
9     CT    16.500000                       28.043820   
10    NH    16.400000                       30.405103   
11    UT    16.333333                       22.130135   
12    NV    16.000000                       25.360870   
13    NJ    15.476190                       25.310135   
14    AZ    15.400000                       21.618954   
15    OH    15.363636                       31.208670   
16    PA    15.230769                       29.900274   
17    FL    14.818182                       19.823734   
18    CA    14.631579                       16.494877   
19    OK    14.492754                       23.194731   
20    NC    13.979798                       21.772484   
21    WA    13.722222                       15.813231   
22    ME    13.250000                       23.628730   
23    OR    13.034483                       13.356529   
24    IN    12.912088                       25.007729   
25    LA    12.566667                       20.998437   
26    VT    12.076923                       19.278777   
27    CO    12.000000                       20.936160   
28    MI    11.780488                       20.295450   
29    MT    11.625000                       13.738386   
30    MO    11.049505                       20.497453   
31    VA    11.000000                       19.951126   
32    AK    10.933333                       20.707149   
33    HI    10.750000                       12.410485   
34    SC    10.377778                       16.803245   
35    WY    10.352941                       18.316737   
36    GA     9.727941                       16.480841   
37    IL     9.540230                       19.211123   
38    NY     9.147541                       18.014534   
39    AR     9.095238                       17.405743   
40    AL     9.083333                       15.289231   
41    ID     8.965517                       17.070726   
42    WI     8.602941                       15.864789   
43    TX     8.335227                       11.464450   
44    KS     8.136364                       13.463310   
45    MS     7.971429                       14.980853   
46    MN     6.824561                       12.458245   
47    IA     5.510204                       11.960783   
48    ND     5.500000                       11.093851   
49    SD     5.071429                        9.420554   
50    NE     4.950000                        7.486563   

    Premature death raw value  Injury deaths raw value  \
0                10389.013702               110.558261   
1                 8132.374322                72.389158   
2                 9935.828031               116.532774   
3                10743.241235                99.606908   
4                 7797.469824                72.947589   
5                 7663.613201                74.908997   
6                10247.637329                96.872238   
7                 5990.369088                62.346762   
8                 5551.636011                74.153465   
9                 5695.160844                68.237936   
10                6783.899091                85.594775   
11                6946.338383                85.088068   
12                8356.648988                96.158885   
13                6324.415440                58.306088   
14                9152.913653               103.702636   
15                8515.141529                80.305389   
16                7649.411904                82.9935

#### 1. Test for Injury deaths raw value : 
    
    H0: The mean of Norm deaths and Injury deaths are the same 
    H1 : The mean of samples is different.

In [8]:
res = stats.ttest_ind(a= df_mean_variables['Injury deaths raw value'],
                b= df_mean_variables['Norm_deaths'],
                equal_var=False)
res

Ttest_indResult(statistic=35.50316801630451, pvalue=5.858361181618531e-42)

In [9]:
if res[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


The null hypothesis is rejected as the the mean of norm deaths and that of injury deaths are not same. Thus we accept the alternative hypothesis. 

#### 2. Test for Drug overdose deaths raw value: 

    H0: The mean of Norm deaths and Drug overdose deaths raw value are the same. 
    H1: The mean of samples is different.

In [10]:
res1= stats.ttest_ind(a = df_mean_variables['Drug overdose deaths raw value'],
                b= df_mean_variables['Norm_deaths'],
                equal_var=False)
res1

Ttest_indResult(statistic=6.845021189563974, pvalue=1.3964724215485585e-09)

In [11]:
if res1[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


The null hypothesis is rejected as the the mean of norm deaths and that of drug overdose deaths raw value are not same. Thus we accept the alternative hypothesis.

#### 3. Test for Premature death raw value value:

     H0: The mean of Norm deaths and  Premature death raw value are same. 
     H1: The mean of samples is different.
    

In [12]:
res2 = stats.ttest_ind(a = df_mean_variables['Premature death raw value'],
                b= df_mean_variables['Norm_deaths'],
                equal_var=False)
res2

Ttest_indResult(statistic=37.34438671711174, pvalue=3.449558797122838e-38)

In [13]:
if res2[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


The null hypothesis is rejected as the mean of norm deaths and that of Premature death raw value are not same. Thus we accept the alternative hypothesis.

#### 4. Test for Insufficient sleep raw value:

     H0: The mean of Norm deaths and  Insufficient sleep raw value are same. 
     H1: The mean of samples is different.

In [14]:
res3 = stats.ttest_ind(a = df_mean_variables['Insufficient sleep raw value'],
                b=df_mean_variables['Norm_deaths'],
                equal_var=False)
res3

Ttest_indResult(statistic=-20.69615094941355, pvalue=3.5691959192290764e-26)

In [15]:
if res3[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


The null hypothesis is rejected as the the mean of norm deaths and that of Insufficient sleep raw value are not same. Thus we accept the alternative hypothesis.

#### 5. Test for Firearm fatalities raw value:

     H0: The mean of Norm deaths and Firearm fatalities raw value are same. 
     H1: The mean of samples is different.

In [16]:
res4 = stats.ttest_ind(a = df_mean_variables['Firearm fatalities raw value'],
                b=df_mean_variables['Norm_deaths'],
                equal_var=False)
res4

Ttest_indResult(statistic=1.593095843741338, pvalue=0.11440157328546836)

In [17]:
if res4[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Accept the Null Hypothesis


#### Almost the mean of both the samples are same and this led to the acceptance of Null hypothesis. 

In [18]:
print(df_mean_variables['Norm_deaths'].mean())
print(df_mean_variables['Firearm fatalities raw value'].mean())

12.724870891553623
14.218496226850425


The null hypothesis is accepted as the mean of norm deaths and that of  Firearm fatalities raw value are. Thus we accept the null hypothesis.

### Another way of performing the null hypothesis

In [19]:
#create a df with the variables of interest
hypo_df = super_df[['County' ,'Norm_deaths' ,'Drug overdose deaths raw value', 'Premature death raw value', 'Injury deaths raw value', 'Adult smoking raw value' , 'Firearm fatalities raw value' , 'Insufficient sleep raw value']]

In [20]:
hypo_df

County  Norm_deaths  \
0              Abbeville County, SC          7.0   
1                 Acadia Parish, LA         19.0   
2               Accomack County, VA         11.0   
3                    Ada County, ID         12.0   
4                  Adair County, KY         11.0   
...                             ...          ...   
2522                Yuba County, CA         11.0   
2523  Yukon-Koyukuk Census Area, AK         11.0   
2524                Yuma County, AZ         12.0   
2525              Zapata County, TX          6.0   
2526              Zavala County, TX          6.0   

      Drug overdose deaths raw value  Premature death raw value  \
0                          17.443577                9517.898548   
1                          13.843631               10194.322670   
2                          11.171482               10459.767290   
3                          13.706962                5027.391209   
4                                NaN                9140.796987   
...                              ...                        ...   
2522                       19.400524                9470.154220   
2523                             NaN               19307.389470   
2524                       15.871988                5834.212268   
2525                             NaN                6770.768596   
2526                             NaN                9389.759155   

      Injury deaths raw value  Adult smoking raw value  \
0                   83.535478                 0.174334   
1                   79.039494                 0.210474   
2                   82.000073                 0.164887   
3                   59.693597                 0.117813   
4                   88.793248                 0.232604   
...                       ...                      ...   
2522                86.339627                 0.148560   
2523               242.174510                 0.313589   
2524                62.042024                 0.148959   
2525                36.234914                 0.179719   
2526                65.975028                 0.196905   

      Firearm fatalities raw value  Insufficient sleep raw value  
0                             23.3                      0.360838  
1                             21.1                      0.323577  
2                             18.2                      0.367502  
3                             11.7                      0.262969  
4                             11.5                      0.364091  
...                            ...                           ...  
2522                          13.9                      0.337518  
2523                          39.8                      0.345312  
2524                          10.7                      0.297741  
2525                           NaN                      0.346546  
2526                           NaN                      0.349164  

[2527 rows x 8 columns]

#### Calculate the mean for norm deaths and divide the dataset into high and low as we are comparing the independant variables with dependant variable

In [21]:
#calculate the mim max and mean to divide the dataset using pd.cut()
mean = hypo_df.Norm_deaths.median()
min = hypo_df.Norm_deaths.min()
max = hypo_df.Norm_deaths.max()
print(mean)
print(min)
print(max)

11.0
1.0
63.0


In [22]:
hypo_df = hypo_df.fillna(0)

In [23]:
#split the data based on the min max and mean values 
hypo_df['death_category'] = pd.cut(super_df.Norm_deaths, [0.99, 16, 64] , labels=['low' , 'high'])


In [24]:
hypo_df

County  Norm_deaths  \
0              Abbeville County, SC          7.0   
1                 Acadia Parish, LA         19.0   
2               Accomack County, VA         11.0   
3                    Ada County, ID         12.0   
4                  Adair County, KY         11.0   
...                             ...          ...   
2522                Yuba County, CA         11.0   
2523  Yukon-Koyukuk Census Area, AK         11.0   
2524                Yuma County, AZ         12.0   
2525              Zapata County, TX          6.0   
2526              Zavala County, TX          6.0   

      Drug overdose deaths raw value  Premature death raw value  \
0                          17.443577                9517.898548   
1                          13.843631               10194.322670   
2                          11.171482               10459.767290   
3                          13.706962                5027.391209   
4                           0.000000                9140.796987   
...                              ...                        ...   
2522                       19.400524                9470.154220   
2523                        0.000000               19307.389470   
2524                       15.871988                5834.212268   
2525                        0.000000                6770.768596   
2526                        0.000000                9389.759155   

      Injury deaths raw value  Adult smoking raw value  \
0                   83.535478                 0.174334   
1                   79.039494                 0.210474   
2                   82.000073                 0.164887   
3                   59.693597                 0.117813   
4                   88.793248                 0.232604   
...                       ...                      ...   
2522                86.339627                 0.148560   
2523               242.174510                 0.313589   
2524                62.042024                 0.148959   
2525                36.234914                 0.179719   
2526                65.975028                 0.196905   

      Firearm fatalities raw value  Insufficient sleep raw value  \
0                             23.3                      0.360838   
1                             21.1                      0.323577   
2                             18.2                      0.367502   
3                             11.7                      0.262969   
4                             11.5                      0.364091   
...                            ...                           ...   
2522                          13.9                      0.337518   
2523                          39.8                      0.345312   
2524                          10.7                      0.297741   
2525                           0.0                      0.346546   
2526                           0.0                      0.349164   

     death_category  
0               low  
1              high  
2               low  
3               low  
4               low  
...             ...  
2522            low  
2523            low  
2524            low  
2525            low  
2526            low  

[2527 rows x 9 columns]

#### Created a dataframe with low and high death category and now will perform the hypothesis analysis on the dataframe. The type of testing I am doing here is a 2-sample test  as I am comparing independant dataframes and then perform hypothesis on that. 

In [25]:
low = hypo_df[hypo_df['death_category'] == 'low']
high =  hypo_df[hypo_df['death_category'] == 'high']

#### 1. Test for Injury deaths raw value: 

    H0: The mean of Norm deaths and Premature death raw value are the same for low dataframe
    H1 : The mean of samples is different.

In [26]:
a = stats.ttest_rel(a = low['Injury deaths raw value'],
                b= low['Norm_deaths'])
a

Ttest_relResult(statistic=148.23958071245326, pvalue=0.0)

In [27]:
if a[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


#### 2. Test for Drug overdose deaths raw value:

    H0: The mean of Norm deaths and Drug overdose deaths raw value are the same for low dataframe. 
    H1: The mean of samples is different.

In [28]:
b = stats.ttest_ind(a = low['Drug overdose deaths raw value'],
                b= low['Norm_deaths'],
                equal_var=False)
b

Ttest_indResult(statistic=3.914266117426446, pvalue=9.30758074320493e-05)

In [29]:
if b[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


#### 3. Test for Premature death raw value value:

     H0: The mean of Norm deaths and  Premature death raw value are same for value are the same for low dataframe. 
     H1: The mean of samples is different.
    

In [30]:
c = stats.ttest_ind(a = low['Premature death raw value'],
                b= low['Norm_deaths'],
                equal_var=False)
c

Ttest_indResult(statistic=151.56844285252635, pvalue=0.0)

In [31]:
if c[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


#### 4. Test for Insufficient sleep raw value:

     H0: The mean of Norm deaths and  Insufficient sleep raw value are same for high dataset. 
     H1: The mean of samples is different.

In [32]:
d = stats.ttest_ind(a = high['Insufficient sleep raw value'],
                b= high['Norm_deaths'],
                equal_var=False)
d

Ttest_indResult(statistic=-80.3730692111519, pvalue=1.1320973193261756e-298)

In [33]:
if d[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


#### 5. Test for Firearm fatalities raw value:

     H0: The mean of Norm deaths and Firearm fatalities raw value are same for high dataset. 
     H1: The mean of samples is different.

In [34]:
e = stats.ttest_ind(a = high['Firearm fatalities raw value'],
                b= high['Norm_deaths'],
                equal_var=False)
e

Ttest_indResult(statistic=-16.548312345459063, pvalue=1.0067793046335344e-54)

In [35]:
if e[1] <= 0.05:
    print("Reject the Null Hypothesis as the pvalue is less than 0.05")
else:
    print("Accept the Null Hypothesis")

Reject the Null Hypothesis as the pvalue is less than 0.05


### Inference: 


* For all the variables the threshold for comparing used is pvalue which tells the probability of seeing a result as extreme or   more extreme than the one observed.  P-value is the probability of obtaining the observed sample results by chance.


* The distribution for both the dataset is not the same and hence we were able to prove by using the 2-sample test for the two dataframes.

### M2.2 Perform linear regression to discover patterns (25 pts)

#### Perform a linear regression between Normalized Mortality and Opiod_Dispensing_Rate

#### We perform linear regression to understand the dependancy of a variable with other independant variable. 

In [36]:
#normalize the opiod dispensing rate 
super_df['Opiod_Dispensing_Rate_Norm'] = (super_df['Opiod_Dispensing_Rate']/super_df['Population'] * 100000).round()

In [37]:
linear_data = super_df[['Opiod_Dispensing_Rate_Norm' , 'Norm_deaths']]

In [38]:
linear_data

Opiod_Dispensing_Rate_Norm  Norm_deaths
0                            6.0          7.0
1                            3.0         19.0
2                            3.0         11.0
3                            1.0         12.0
4                           18.0         11.0
...                          ...          ...
2522                         2.0         11.0
2523                         1.0         11.0
2524                         1.0         12.0
2525                         1.0          6.0
2526                         1.0          6.0

[2527 rows x 2 columns]

In [39]:
# create a fitted model using smf.ols
lm = smf.ols(formula='Norm_deaths ~ Opiod_Dispensing_Rate_Norm', data=linear_data).fit()

In [40]:
preds = lm.predict(linear_data)

In [41]:
lm.params

Intercept                     12.037933
Opiod_Dispensing_Rate_Norm     0.015193
dtype: float64

In [42]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            Norm_deaths   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     3.691
Date:                Mon, 21 Nov 2022   Prob (F-statistic):             0.0548
Time:                        21:47:09   Log-Likelihood:                -8415.5
No. Observations:                2527   AIC:                         1.683e+04
Df Residuals:                    2525   BIC:                         1.685e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept                     12.0379      0.146     82.309      0.000      11.751      12.325
Opiod_Dispensing_Rate_Norm     0.0152      0.008      1.921      0.055      -0.000       0.031
==============================================================================
Omnibus:                      888.922   Durbin-Watson:                   1.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4162.115
Skew:                           1.627   Prob(JB):                         0.00
Kurtosis:                       8.379   Cond. No.                         20.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Inference : 

* The independent variable is opid dispensing rate and the dependent variable is norm deaths. 
*  The R-squared value is 0.001 i.e with just one independent variable — accounts for roughly 1% of the variance in death rates 
* we see most of the deaths were when the  opiod dispensing rates was equal to 100 units. 
* From the above parameters we see that the slope is positive this means that with the increase in opiod dispensing rate,        the deaths also increased.
* We can interpret the coefficient of opiod dispensing rate as, every year the deaths for opiod dispensing rate is found to increase by 12%
 

## Multiple linear regression with 5 variables

In [43]:
#renaming the column names for simplicity
new_df = super_df.rename(columns={"Drug overdose deaths raw value": "Drug_overdose_deaths_raw_value", "Injury deaths raw value": "Injury_deaths_raw_value", "Adult smoking raw value": "Adult_smoking_raw_value" ,"Firearm fatalities raw value": "Firearm_fatalities_raw_value" , "Insufficient sleep raw value" : "Insufficient_sleep_raw_value"})

#### We need to normalize the variable that we are comparing to population as Opiod_dispensing rate is normalised. Thus we normalize all the variables and then perform the multiple regression

In [44]:
new_df['Opiod_Dispensing_Rate_Norm'] = (new_df['Opiod_Dispensing_Rate']/super_df['Population'] * 100000).round()

In [45]:
new_df['Drug_overdose_deaths_raw_value_Norm'] = (new_df['Drug_overdose_deaths_raw_value']/super_df['Population'] * 100000).round()
new_df['Injury_deaths_raw_value_Norm'] = (new_df['Injury_deaths_raw_value']/super_df['Population'] * 100000).round()
new_df['Adult_smoking_raw_value_Norm'] = (new_df['Adult_smoking_raw_value']/super_df['Population'] * 100000).round()
new_df['Firearm_fatalities_raw_value_Norm'] = (new_df['Firearm_fatalities_raw_value']/super_df['Population'] * 100000).round()
new_df['Insufficient_sleep_raw_value_Norm'] = (new_df['Insufficient_sleep_raw_value']/super_df['Population'] * 100000).round()

In [46]:
#filling the nan values with zero after normalization
new_df = new_df.fillna(0)

In [47]:
# create a fitted model with all the features of interest
lm = smf.ols(formula='Opiod_Dispensing_Rate_Norm ~ Drug_overdose_deaths_raw_value_Norm + Injury_deaths_raw_value_Norm + Adult_smoking_raw_value_Norm + Firearm_fatalities_raw_value_Norm + Insufficient_sleep_raw_value_Norm', data=new_df).fit()

# print the coefficients
lm.params

Intercept                              2.112730
Drug_overdose_deaths_raw_value_Norm    0.077753
Injury_deaths_raw_value_Norm           0.325279
Adult_smoking_raw_value_Norm           0.000000
Firearm_fatalities_raw_value_Norm     -0.281828
Insufficient_sleep_raw_value_Norm      0.000000
dtype: float64

In [48]:
lm.summary()

C:\Users\p_bud\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1918: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])


<class 'statsmodels.iolib.summary.Summary'>
"""
                                OLS Regression Results                                
======================================================================================
Dep. Variable:     Opiod_Dispensing_Rate_Norm   R-squared:                       0.146
Model:                                    OLS   Adj. R-squared:                  0.145
Method:                         Least Squares   F-statistic:                     143.7
Date:                        Mon, 21 Nov 2022   Prob (F-statistic):           6.14e-86
Time:                                21:47:09   Log-Likelihood:                -10548.
No. Observations:                        2527   AIC:                         2.110e+04
Df Residuals:                            2523   BIC:                         2.113e+04
Df Model:                                   3                                         
Covariance Type:                    nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               2.1127      0.431      4.905      0.000       1.268       2.957
Drug_overdose_deaths_raw_value_Norm     0.0778      0.099      0.786      0.432      -0.116       0.272
Injury_deaths_raw_value_Norm            0.3253      0.017     19.231      0.000       0.292       0.358
Adult_smoking_raw_value_Norm                 0          0        nan        nan           0           0
Firearm_fatalities_raw_value_Norm      -0.2818      0.104     -2.705      0.007      -0.486      -0.078
Insufficient_sleep_raw_value_Norm            0          0        nan        nan           0           0
==============================================================================
Omnibus:                     6615.215   Durbin-Watson:                   2.012
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        152840301.847
Skew:                          29.602   Prob(JB):                         0.00
Kurtosis:                    1206.364   Cond. No.                          inf
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#### Inference : 

* Here the opioid dispensing rate is the dependent variable, and the rest other variables are independent.
* Injury_deaths_raw_value is  0.325279 positively related , Firearm_fatalities_raw_value is -0.281828 are negatively related. So when the slope is positive if independant variable increases by value then the dependent variable will also increase.

* Adult_smoking_raw_value_Norm , Insufficient_sleep_raw_value_Norm are 0 so that means the slope is zeo and the variables are not related. 

* Drug_overdose_deaths_raw_value is 0.0778 this means the slope is positive and the hence the death increase with every 7.78% increase in drug overdose. 

* The R-squared is 0.146 for this model and with 5 independent variables — accounts for roughly 14% of the variance in opioid dispensing rates.

#### Test non-linear model with the 5 + 1 variables (n=2,3,4)

#### Since there are 5 variables and the degree of 0 is 1 , I added 1 to all the variables. 

In [49]:
new_df.head(3)

Unnamed: 0                County  County Code  Population  Deaths  \
0           0  Abbeville County, SC        45001      535389      38   
1           1     Acadia Parish, LA        22001     1279727     237   
2           2   Accomack County, VA        51001      726189      73   

   Norm_deaths  State FIPS Code  County FIPS Code  5-digit FIPS Code  \
0          7.0               45                 1              45001   
1         19.0               22                 1              22001   
2         11.0               51                 1              51001   

  State Abbreviation  ... Population growth raw value  State   FIPS  \
0                 SC  ...                         0.0     SC  45001   
1                 LA  ...                         0.0     LA  22001   
2                 VA  ...                         0.0     VA  51001   

   Opiod_Dispensing_Rate  Opiod_Dispensing_Rate_Norm  \
0                   32.4                         6.0   
1                   32.2                         3.0   
2                   19.4                         3.0   

   Drug_overdose_deaths_raw_value_Norm  Injury_deaths_raw_value_Norm  \
0                                  3.0                          16.0   
1                                  1.0                           6.0   
2                                  2.0                          11.0   

   Adult_smoking_raw_value_Norm  Firearm_fatalities_raw_value_Norm  \
0                           0.0                                4.0   
1                           0.0                                2.0   
2                           0.0                                3.0   

   Insufficient_sleep_raw_value_Norm  
0                                0.0  
1                                0.0  
2                                0.0  

[3 rows x 549 columns]

In [50]:
x = new_df[['Opiod_Dispensing_Rate_Norm' ,'Drug_overdose_deaths_raw_value_Norm', 'Injury_deaths_raw_value_Norm', 'Adult_smoking_raw_value_Norm','Firearm_fatalities_raw_value_Norm','Insufficient_sleep_raw_value_Norm']]

In [51]:
# 2-nd order polynomial
poly_2 = smf.ols(formula='Opiod_Dispensing_Rate_Norm ~ 1 + Drug_overdose_deaths_raw_value_Norm + 1 + Firearm_fatalities_raw_value_Norm + 1 + Injury_deaths_raw_value_Norm + 1 + Adult_smoking_raw_value_Norm+ Insufficient_sleep_raw_value_Norm + 1 + I(Drug_overdose_deaths_raw_value_Norm ** 2.0) + I(Firearm_fatalities_raw_value_Norm ** 2) +  I(Injury_deaths_raw_value_Norm **2) + I(Adult_smoking_raw_value_Norm **2) + I(Insufficient_sleep_raw_value_Norm **2)', data=x).fit()

In [53]:
poly_2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                OLS Regression Results                                
======================================================================================
Dep. Variable:     Opiod_Dispensing_Rate_Norm   R-squared:                       0.157
Model:                                    OLS   Adj. R-squared:                  0.155
Method:                         Least Squares   F-statistic:                     78.31
Date:                        Mon, 21 Nov 2022   Prob (F-statistic):           5.51e-90
Time:                                21:47:09   Log-Likelihood:                -10532.
No. Observations:                        2527   AIC:                         2.108e+04
Df Residuals:                            2520   BIC:                         2.112e+04
Df Model:                                   6                                         
Covariance Type:                    nonrobust                                         
=================================================================================================================
                                                    coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------------
Intercept                                         0.7354      0.503      1.461      0.144      -0.252       1.723
Drug_overdose_deaths_raw_value_Norm               0.0915      0.168      0.545      0.586      -0.238       0.421
Firearm_fatalities_raw_value_Norm                -0.4036      0.161     -2.502      0.012      -0.720      -0.087
Injury_deaths_raw_value_Norm                      0.4829      0.034     14.238      0.000       0.416       0.549
Adult_smoking_raw_value_Norm                  -5.626e-16   6.33e-16     -0.889      0.374    -1.8e-15    6.79e-16
Insufficient_sleep_raw_value_Norm             -1.945e-16   1.07e-16     -1.815      0.070   -4.05e-16    1.56e-17
I(Drug_overdose_deaths_raw_value_Norm ** 2.0)    -0.0066      0.007     -0.912      0.362      -0.021       0.008
I(Firearm_fatalities_raw_value_Norm ** 2)        -0.0008      0.005     -0.161      0.872      -0.011       0.009
I(Injury_deaths_raw_value_Norm ** 2)             -0.0014      0.000     -5.222      0.000      -0.002      -0.001
I(Adult_smoking_raw_value_Norm ** 2)                   0          0        nan        nan           0           0
I(Insufficient_sleep_raw_value_Norm ** 2)              0          0        nan        nan           0           0
==============================================================================
Omnibus:                     6652.600   Durbin-Watson:                   2.024
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        157941951.188
Skew:                          30.037   Prob(JB):                         0.00
Kurtosis:                    1226.288   Cond. No.                          inf
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#### Inference: 

From the summary table we conclude the below: 

* R -square is the coefficient of determination that tells us that how much percentage variation dependent variable can be    explained by independent variable. Here, 15.7 % variation in Opiod dispemsing rate can be explained by the 5 variables that we have chosen. The maximum R-squared value can be 1 , which means that the larger the R-squared value better is the regression.

* The coefficient term tells the change in opiod dispensing rate for a unit change in independant variables i.e if independant variables rises by 1 unit then opiod dispensing rate rises by 9.15% for drug overdose , 40.36% for firearm fatalities and as its negatively related there is no dependancy on opiod dispensing rate , 48% injury deaths increase with increase in opiod rate, and since adult smoking , insufficient sleep are negatively related to opiod dispensing rate as the slope is negative. 
    

In [54]:
# 3-rd order polynomial
poly_3 = smf.ols(formula='Opiod_Dispensing_Rate_Norm ~ 1 + Drug_overdose_deaths_raw_value_Norm + Firearm_fatalities_raw_value_Norm + Injury_deaths_raw_value_Norm + Adult_smoking_raw_value_Norm + Insufficient_sleep_raw_value_Norm + I(Drug_overdose_deaths_raw_value_Norm ** 2.0) + I(Firearm_fatalities_raw_value_Norm ** 2) + I(Injury_deaths_raw_value_Norm **2) + I(Adult_smoking_raw_value_Norm **2) + I(Insufficient_sleep_raw_value_Norm **2) + I(Drug_overdose_deaths_raw_value_Norm ** 3.0) + I(Firearm_fatalities_raw_value_Norm ** 3) + I(Injury_deaths_raw_value_Norm **3) + I(Adult_smoking_raw_value_Norm **3) + I(Insufficient_sleep_raw_value_Norm **3) ', data=x).fit()

In [55]:
poly_3.summary()

C:\Users\p_bud\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1918: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])


<class 'statsmodels.iolib.summary.Summary'>
"""
                                OLS Regression Results                                
======================================================================================
Dep. Variable:     Opiod_Dispensing_Rate_Norm   R-squared:                       0.163
Model:                                    OLS   Adj. R-squared:                  0.160
Method:                         Least Squares   F-statistic:                     54.56
Date:                        Mon, 21 Nov 2022   Prob (F-statistic):           4.22e-91
Time:                                21:47:09   Log-Likelihood:                -10522.
No. Observations:                        2527   AIC:                         2.106e+04
Df Residuals:                            2517   BIC:                         2.112e+04
Df Model:                                   9                                         
Covariance Type:                    nonrobust                                         
=================================================================================================================
                                                    coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------------
Intercept                                         0.8665      0.584      1.483      0.138      -0.279       2.012
Drug_overdose_deaths_raw_value_Norm               0.4898      0.289      1.696      0.090      -0.077       1.056
Firearm_fatalities_raw_value_Norm                 0.3164      0.288      1.097      0.273      -0.249       0.882
Injury_deaths_raw_value_Norm                      0.2992      0.062      4.792      0.000       0.177       0.422
Adult_smoking_raw_value_Norm                  -6.799e-14   4.19e-14     -1.622      0.105    -1.5e-13    1.42e-14
Insufficient_sleep_raw_value_Norm             -2.475e-15   1.24e-15     -1.989      0.047   -4.92e-15    -3.5e-17
I(Drug_overdose_deaths_raw_value_Norm ** 2.0)    -0.0524      0.029     -1.834      0.067      -0.108       0.004
I(Firearm_fatalities_raw_value_Norm ** 2)        -0.0622      0.023     -2.735      0.006      -0.107      -0.018
I(Injury_deaths_raw_value_Norm ** 2)              0.0026      0.001      2.343      0.019       0.000       0.005
I(Adult_smoking_raw_value_Norm ** 2)           1.462e-17   7.04e-17      0.208      0.836   -1.23e-16    1.53e-16
I(Insufficient_sleep_raw_value_Norm ** 2)              0          0        nan        nan           0           0
I(Drug_overdose_deaths_raw_value_Norm ** 3.0)     0.0009      0.001      1.593      0.111      -0.000       0.002
I(Firearm_fatalities_raw_value_Norm ** 3)         0.0008      0.000      2.617      0.009       0.000       0.001
I(Injury_deaths_raw_value_Norm ** 3)          -1.562e-05   4.27e-06     -3.657      0.000    -2.4e-05   -7.24e-06
I(Adult_smoking_raw_value_Norm ** 3)                   0          0        nan        nan           0           0
I(Insufficient_sleep_raw_value_Norm ** 3)              0          0        nan        nan           0           0
==============================================================================
Omnibus:                     6622.534   Durbin-Watson:                   2.022
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        153833476.425
Skew:                          29.687   Prob(JB):                         0.00
Kurtosis:                    1210.268   Cond. No.                          inf
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

### Inference

* R-Square is the coefficient of determination that tells us that how much percentage variation dependent variable can be    explained by independent variable. Here, 16.3 % variation in Opiod dispemsing rate can be explained by the 5 variables that we have chosen. The maximum R-squared value can be 1, which means that the larger the R-squared value better is the regression.


* The coefficient term tells the change in opiod dispending rate for a unit change in independant variables i.e if independant variables rises by 1 unit then opiod dispensing rate rises by 48.9% for drug overdose , 0.31% for firearm fatalities , 29.92% injury deaths, adult smoking and insufficient sleep are negatively correlated. 
    

In [56]:
#4-th order polynomial
poly_4 = smf.ols(formula='Opiod_Dispensing_Rate_Norm ~ 1 + Drug_overdose_deaths_raw_value_Norm + 1 + Firearm_fatalities_raw_value_Norm + 1 + Injury_deaths_raw_value_Norm + 1 + Adult_smoking_raw_value_Norm+ 1 + Insufficient_sleep_raw_value_Norm + I(Drug_overdose_deaths_raw_value_Norm ** 2.0) + I(Firearm_fatalities_raw_value_Norm ** 2) + I(Injury_deaths_raw_value_Norm **2) + I(Adult_smoking_raw_value_Norm **2) + I(Insufficient_sleep_raw_value_Norm **2) + I(Drug_overdose_deaths_raw_value_Norm ** 3.0) + I(Firearm_fatalities_raw_value_Norm ** 3) + I(Injury_deaths_raw_value_Norm **3) + I(Adult_smoking_raw_value_Norm **3) + I(Insufficient_sleep_raw_value_Norm **3)+ I(Drug_overdose_deaths_raw_value_Norm ** 4.0) + I(Firearm_fatalities_raw_value_Norm **4) + I(Injury_deaths_raw_value_Norm **4) + I(Adult_smoking_raw_value_Norm **4) + I(Insufficient_sleep_raw_value_Norm **4)', data=x).fit()

In [57]:
poly_4.summary()

C:\Users\p_bud\anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1918: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])


<class 'statsmodels.iolib.summary.Summary'>
"""
                                OLS Regression Results                                
======================================================================================
Dep. Variable:     Opiod_Dispensing_Rate_Norm   R-squared:                       0.167
Model:                                    OLS   Adj. R-squared:                  0.163
Method:                         Least Squares   F-statistic:                     41.92
Date:                        Mon, 21 Nov 2022   Prob (F-statistic):           9.47e-91
Time:                                21:47:10   Log-Likelihood:                -10517.
No. Observations:                        2527   AIC:                         2.106e+04
Df Residuals:                            2514   BIC:                         2.114e+04
Df Model:                                  12                                         
Covariance Type:                    nonrobust                                         
=================================================================================================================
                                                    coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------------
Intercept                                         1.2266      0.657      1.867      0.062      -0.062       2.515
Drug_overdose_deaths_raw_value_Norm               0.4663      0.429      1.086      0.278      -0.376       1.308
Firearm_fatalities_raw_value_Norm                 1.4465      0.487      2.967      0.003       0.491       2.402
Injury_deaths_raw_value_Norm                      0.0508      0.105      0.482      0.630      -0.156       0.258
Adult_smoking_raw_value_Norm                   4.748e-12   6.18e-12      0.768      0.443   -7.38e-12    1.69e-11
Insufficient_sleep_raw_value_Norm             -9.862e-14   4.65e-14     -2.123      0.034    -1.9e-13   -7.54e-15
I(Drug_overdose_deaths_raw_value_Norm ** 2.0)    -0.0469      0.072     -0.647      0.518      -0.189       0.095
I(Firearm_fatalities_raw_value_Norm ** 2)        -0.2508      0.073     -3.414      0.001      -0.395      -0.107
I(Injury_deaths_raw_value_Norm ** 2)              0.0112      0.003      3.606      0.000       0.005       0.017
I(Adult_smoking_raw_value_Norm ** 2)          -1.182e-16   3.39e-16     -0.349      0.727   -7.83e-16    5.46e-16
I(Insufficient_sleep_raw_value_Norm ** 2)       2.07e-15    1.3e-15      1.593      0.111   -4.77e-16    4.62e-15
I(Drug_overdose_deaths_raw_value_Norm ** 3.0)     0.0004      0.004      0.107      0.914      -0.007       0.007
I(Firearm_fatalities_raw_value_Norm ** 3)         0.0084      0.003      2.828      0.005       0.003       0.014
I(Injury_deaths_raw_value_Norm ** 3)          -9.593e-05   2.75e-05     -3.487      0.000      -0.000    -4.2e-05
I(Adult_smoking_raw_value_Norm ** 3)                   0          0        nan        nan           0           0
I(Insufficient_sleep_raw_value_Norm ** 3)              0          0        nan        nan           0           0
I(Drug_overdose_deaths_raw_value_Norm ** 4.0)  7.558e-06   4.75e-05      0.159      0.874   -8.56e-05       0.000
I(Firearm_fatalities_raw_value_Norm ** 4)     -7.636e-05   3.04e-05     -2.508      0.012      -0.000   -1.67e-05
I(Injury_deaths_raw_value_Norm ** 4)           2.102e-07   7.17e-08      2.931      0.003    6.96e-08    3.51e-07
I(Adult_smoking_raw_value_Norm ** 4)                   0          0        nan        nan           0           0
I(Insufficient_sleep_raw_value_Norm ** 4)              0          0        nan        nan           0           0
==============================================================================
Omnibus:                     6603.683   Durbin-Watson:                   2.021
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        150449549.606
Skew:                 

### Inference

* R2 is the coefficient of determination that tells us that how much percentage variation dependent variable can be    explained by independent variable. Here, 16.7 % variation in Opiod dispemsing rate can be explained by the 5 variables that we have chosen. The maximum R-squared value can be 1, which means that the larger the R-squared value better is the regression.


* The coefficient term tells the change in opiod dispending rate for a unit change in independant variables i.e if independant variables rises by 1 unit then opiod dispending rate rises by 46.63% for drug overdose , 144.6% for firearm fatalities , approx 5% injury deaths, aprox 0 % for adult smoking i.e dependant and independant variables are not related and approx 0 %  in insufficient sleep as the dependant and independant variables are not related since the slopes are negative.
    

#### Thus from the R-Square value I can infer that by increasing the polynomial degree there wasnt much significant difference in the models. Model with degree 3 and degree 4 behaved almost the same and is better than degree 2 since the R-square values for those model were better. Also, when the polynomial degree increases, the model performs better and there is possibility of overfitting the data.

#### References: 
    
    * https://github.com/UNCG-CSE/CSC-605_Fall_2022/blob/main/Class_Resources/Lecture_06/Statistics/05_Hypothesis.ipynb
    
    * https://github.com/UNCG-CSE/CSC-605_Fall_2022/blob/main/Class_Resources/Lecture_06/Statistics/07_linear_regression.ipynb 